keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,703 @@
+notes right away. Also the kubeadm maintainers might make mistakes forgetting
+about a certain feature/bug fix.
+
+### Test Plan
+
+#### CI artifacts
+
+Currently the kubeadm binary is built and uploaded in the CI GCS bucket
+next to the rest of the k8sbinaries like kubelet.
+
+There is a complication as now another repository has to be monitored
+for changes. Currently the CI build happens after a PR merged in the
+k/k repository.
+
+The proposal is to always rebuild kubeadm based on the latest tips of
+the master and release-x.yy branches of the k/kubeadm repository after
+a CI build for the k/k repository triggers. In this case the k/release
+build tooling must not trigger the writing of a KUBE_VERSION file in the
+k/kubeadm repository but rather let the version to generate based on the
+latest Git tags and SHAs.
+
+#### PR testing
+
+There is already support for running pre-submit jobs on k/kubeadm PRs.
+Once kubeadm is moved to the new location a collection of PR blocking
+jobs should be added to make sure the project is always buildable with
+the current version of Go that the k/k repository uses. Syncing Go versions
+should be the responsibility of the k/kubeadm maintainers.
+
+The new PR blocking jobs must support building kubeadm from source with
+the latest commits that a PR for k/kubeadm introduces. The rest of the
+k8s artifacts can be pulled from the CI GCS bucket. This can also allow
+catching problems introduced in k/k when using kubeadm as the deployer.
+
+The /hack/verify-* and update-* scripts that k/kubeadm/kubeadm uses
+will not be a direct copy of k/k but will provide similar results:
+Go versions are in sync Go-fmt, linter and import tools are in sync
+as much as possible
+
+The following lists of tests will be executed on PRs as blocking:
+- Unit tests
+- Integration tests (currently reside in k/k/cmd/kubeadm/test)
+- E2e tests from the e2e-kubeadm suite (currently resides
+in k/k/test/e2e-kubeadm)
+- Full cluster deployment e2e, possibly upgrade and skew tests too
+
+#### Periodics jobs
+
+No changes to the current kubeadm periodic jobs will be required
+as long as the CI build process continues to push the kubeadm
+and the rest of the k8s binaries to the GCS buckets.
+
+### Graduation Criteria
+
+In the 1.18 cycle there could be partial work related to copying the
+kubeadm source code to the k/kubeadm repository and experimenting
+with build tooling and the post-submit job.
+
+The finalization of the move is planned for the release of 1.19.
+
+### Upgrade / Downgrade Strategy
+
+No changes to the upgrade strategy will be required.
+Downgrades are not supported by kubeadm.
+
+### Version Skew Strategy
+
+No changes will be applied to the current version skew support in kubeadm
+as long as kubeadm preserves its existing versioning, which matches
+the versioning of k8s.
+
+## Implementation History
+
+- 2019-12-29: This proposal document was created
+
+## Open Questions
```

Comment:
How do we integrate back against testing kubernetes? Do we use kubeadm from HEAD? Do we have a green build?
How do tools consume this?
What is the test matrix?

Reply:
> How do we integrate back against testing kubernetes? Do we use kubeadm from HEAD?

kubeadm has to be built from HEAD for CI.
with this proposal kubeadm will be build from HEAD for release too, but with coordination from the maintainers.

> Do we have a green build?

i'd argue that currently in k8s we have a good concept of what a green build is. we often release a kubelet with a critical bug, but due to not sufficient test coverage we don't catch it. `1.17.0` was not a green build for the kube-apiserver.

for kubeadm will have PR blocking jobs that ensure the project is buildable and passes cluster creation tests.

> How do tools consume this?

if we accept the `make release` proposal, it will be possible to build kubeadm from k/k and one can consume it from the k/k output folders. existing tooling will continue to upload kubeadm artifacts to the same CI/release GCS locations.

> What is the test matrix?

the kubeadm version will match the version of k/k.

kubeadm PR blocking jobs will test the kubeadm HEAD against the k/k HEAD.
kind PR blocking jobs in k/k should do the same.

existing kubeadm periodic jobs will not see changes.

let met know if you need further details or if you want some of this added in the KEP?